SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp

```diff
@@ -0,0 +1,176 @@
+//==------- operator_bitwise_not.cpp  - DPC++ ESIMD on-device test ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu, level_zero
+// XREQUIRES: gpu
+// TODO gpu and level_zero in REQUIRES due to only this platforms supported yet.
+// The current "REQUIRES" should be replaced with "gpu" only as mentioned in
+// "XREQUIRES".
+// UNSUPPORTED: cuda, hip
+// RUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// Test for simd bitwise not operator.
+// The test creates source simd instance with reference data and invokes bitwise
+// not operator.
+// The test verifies that data from simd is not corrupted after calling bitwise
+// not operator, that bitwise not operator return type is as expected and
+// bitwise not operator result values is correct.
+
+#include "../mutator.hpp"
+#include "../shared_element.hpp"
+#include "common.hpp"
+// For std::abs
+#include <cmath>
+
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling bitwise not operator.
+struct bitwise_not_operator {
+  static std::string get_description() { return "bitwise not"; }
+
+  template <typename DataT, int NumElems>
+  static bool call_operator(const DataT *const ref_data,
+                            DataT *const source_simd_result,
+                            DataT *const operator_result) {
+    auto simd_obj = simd<DataT, NumElems>();
+    simd_obj.copy_from(ref_data);
+    auto bitwise_not_result = ~simd_obj;
+    simd_obj.copy_to(source_simd_result);
+    bitwise_not_result.copy_to(operator_result);
+    return std::is_same_v<decltype(~simd_obj), simd<DataT, NumElems>>;
+  }
+};
+
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `m_value` later.
+template <typename T> class For_bitwise_not {
+  T m_value;
+
+public:
+  For_bitwise_not() = default;
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T>) {
+      if (std::abs(value<T>::lowest() + 1) == (value<T>::max() - 1)) {
```

Comment:
```suggestion
      if (std::abs(value<T>::lowest() + 1) == (value<T>::max() - 1)) {
      // We are not using two's complement signed number representation
```
On the second thought, even C++20 having only two's complement representation still has UB based on real-world compiler optimizations for signed integer value overflow. Could you possibly double-check if this condition is required? My concern is possibility of compiler optimizations in case of max/min usage for two's complement signed number representation.

Reply:
The `For_bitwise_not` has been updated in this commit: c43b67ab29daf8efbe8da054c33421dd1b03050f